nir/src/main/scala/scala/scalanative/nir/serialization/BinarySerializer.scala

```diff
@@ -29,23 +42,23 @@ final class BinarySerializer(buffer: ByteBuffer) {
 
     putSeq(names) { n =>
       putGlobal(n)
-      positions += buffer.position()
+      positions += currentPosition()
       putInt(0)
     }
 
-    val offsets = defns.map { defn =>
-      val pos: Int = buffer.position()
-      putDefn(defn)
-      pos
-    }
-    val end = buffer.position()
+    defns
+      .zip(positions)
+      .foreach {
+        case (defn, marker) =>
+          val offset: Int = currentPosition()
+          bufferUnderyling.jumpTo(marker)
+          putInt(offset)
+          bufferUnderyling.continue()
+          putDefn(defn)
+      }
 
-    positions.zip(offsets).map {
-      case (pos, offset) =>
-        buffer.position(pos)
-        putInt(offset)
-    }
-    buffer.position(end)
+    bufferUnderyling.writeTo(outputStream)
+    buffer.flush()
```

Comment:
These two need to be reversed, otherwise `flush()` is futile:
```suggestion
    buffer.flush()
    bufferUnderyling.writeTo(outputStream)
```